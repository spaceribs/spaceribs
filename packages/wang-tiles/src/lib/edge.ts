/**
 * A unique edge representing if two tiles can be validly placed next to each other.
 */
export type Edge = symbol | string;
