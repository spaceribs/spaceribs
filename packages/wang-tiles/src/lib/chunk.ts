import { WangTile } from './tile';
import { CoordMap } from './coord-map';
import { WangTileSet } from './tile-set';
import { RandomGenerator, WeightedArray } from './prng';
import { Edge } from './edge';

/**
 * A chunk is a block of the dimension which can be used to
 * observe part of a tile generated world.
 */
export class WangTileChunk<
  D extends number[],
  V extends WangTileSet<D> | WangTile<D> = WangTileSet<D> | WangTile<D>
> {
  public observed = false;
  public active: D | null = null;
  public readonly coordMap: CoordMap<D, V>;

  private readonly center: D;

  constructor(
    private readonly tileset: Set<WangTile<D>>,
    private readonly random: RandomGenerator,
    chunkStart: D,
    chunkEnd: D
  ) {
    this.coordMap = new CoordMap<D, V>().fill(
      () => new Set(tileset) as V,
      chunkStart,
      chunkEnd
    );
    this.center = chunkStart.map((start, index) => {
      const end = chunkEnd[index];
      const diff = end - start;
      return start + Math.floor(diff / 2);
    }) as D;
  }

  /**
   * Observe the tiles until there are no unobserved tiles.
   * @yields
   */
  public *observe() {
    const centerTile = this.coordMap.get(this.center);

    // observe the center if still a set
    if (centerTile instanceof Set) {
      yield this.select(this.center);
    }

    // sort by size and observe the next smallest set.
    let nextSet;
    do {
      nextSet = this.findSmallestSet();
      if (nextSet != null) {
        this.active = nextSet[0];
        this.reduceAll();
        yield this.select(nextSet[0]);
      }
    } while (nextSet != null);

    this.active = null;

    // mark the chuck as observed and end iteration.
    this.observed = true;
    return;
  }

  /**
   * Select an element via weighted random at the coordinates.
   * @param coord - The coordinates to retrieve.
   * @returns the coordinate/tile pair that was set.
   */
  private select(coord: D): [D, WangTile<D>] {
    const elem = this.reduce(coord);

    if (elem instanceof WangTile) {
      this.coordMap.set(coord, elem);
      return [coord, elem];
    }

    const possibleElements: WeightedArray<WangTile<D>> = Array.from(elem).map(
      (tile) => [tile.probability, tile]
    );

    const tile = this.random.weightedRandom(possibleElements);
    this.coordMap.set(coord, tile as V);
    return [coord, tile];
  }

  /**
   * Find the smallest set in the coordinate map and return it, or null if no more sets exist.
   * @returns A coord/tileset pair or null
   */
  private findSmallestSet(): [D, WangTileSet<D>] | null {
    const sets: [D, WangTileSet<D>][] = Array.from(this.coordMap).filter(
      (coord) => coord[1] instanceof Set
    ) as [D, WangTileSet<D>][];

    if (sets.length < 1) {
      return null;
    }

    if (sets.length == 1) {
      return sets[0];
    }

    return sets.sort((a, b) => {
      return (
        this.getPossibilitiesCount(a[0]) - this.getPossibilitiesCount(b[0])
      );
    })[0];
  }

  /**
   * Get the number for possibilities for a coordinate via count.
   * @param coord - The coordinate to calculate.
   * @returns the count of possibilities for a coordinate.
   */
  private getPossibilitiesCount(coord: D): number {
    const neighbors = this.coordMap.getNeighbors(coord);

    let possibilities = 0;

    for (let dimension = 0; dimension < coord.length; dimension++) {
      const [[, beforeVal], [, afterVal]] = neighbors[dimension];

      if (beforeVal instanceof WangTile) {
        possibilities -= this.tileset.size;
      } else if (beforeVal instanceof Set) {
        possibilities += beforeVal.size;
      } else if (beforeVal == null) {
        possibilities += this.tileset.size * 2;
      }

      if (afterVal instanceof WangTile) {
        possibilities -= this.tileset.size;
      } else if (afterVal instanceof Set) {
        possibilities += afterVal.size;
      } else if (beforeVal == null) {
        possibilities += this.tileset.size * 2;
      }
    }

    return possibilities;
  }

  /**
   * Reduce the possibilities of the entire chunk
   */
  private reduceAll(): void {
    // for (const [coord] of this.coordMap) {
    //   this.reduce(coord);
    // }
  }

  /**
   * Reduce the possibilities to a single tile.
   * @param coord - The coordinate to retrieve.
   * @returns A single, randomly selected tile from the set.
   */
  private reduce(coord: D): V {
    const elem = this.coordMap.get(coord);

    // Throw if coordinates are uninitialized.
    if (elem == null) {
      throw new Error('Coordinates are uninitialized or out of bounds.');
    }

    // Return if already observed.
    if (elem instanceof WangTile) {
      return elem;
    }

    // Throw if the set is empty.
    if (elem.size <= 0) {
      throw new Error('Tileset possibilities were empty.');
    }

    // Return if the set only has one possibility.
    if (elem.size <= 1) {
      return elem.values().next().value;
    }

    const neighbors = this.coordMap.getNeighbors(coord);

    let availableTiles = new WangTileSet(elem);

    for (let dimension = 0; dimension < coord.length; dimension++) {
      const [[beforeCoord, beforeVal], [afterCoord, afterVal]] =
        neighbors[dimension];

      if (beforeVal instanceof WangTile) {
        availableTiles = this.reduceByNeighborEdge(
          beforeVal,
          availableTiles,
          coord.length,
          dimension,
          1
        );
      } else if (beforeVal instanceof Set) {
        availableTiles = this.reduceByNeighborEdges(
          beforeVal,
          availableTiles,
          coord.length,
          dimension,
          1
        );
      }

      if (afterVal instanceof WangTile) {
        availableTiles = this.reduceByNeighborEdge(
          afterVal,
          availableTiles,
          coord.length,
          dimension,
          -1
        );
      } else if (afterVal instanceof Set) {
        availableTiles = this.reduceByNeighborEdges(
          afterVal,
          availableTiles,
          coord.length,
          dimension,
          -1
        );
      }

      if (availableTiles.size < 1) {
        throw new Error(
          `The coordinate {${coord.join()}} could not be resolved due to it's neighbors constraints at {${beforeCoord.join()}}:${beforeVal} and {${afterCoord.join()}}:${afterVal}`
        );
      }
    }

    if (availableTiles.size > 1) {
      const newTileset = new WangTileSet(availableTiles) as V;
      this.coordMap.set(coord, newTileset);
      return newTileset;
    } else {
      const newTile = availableTiles.values().next().value as V;
      this.coordMap.set(coord, newTile);
      return newTile;
    }
  }

  /**
   * Reduce a tileset by a neighboring tile.
   * @param tile - The neighboring tile
   * @param availableTiles - The tileset to reduce
   * @param dimensions - The number of dimensions
   * @param dimension - The dimension of the edge
   * @param offset - The offset from the position
   * @returns - A new tileset reduced by the neighboring tile.
   */
  private reduceByNeighborEdge(
    tile: WangTile<D, unknown>,
    availableTiles: WangTileSet<D, unknown>,
    dimensions: number,
    dimension: number,
    offset: -1 | 1
  ): WangTileSet<D, unknown> {
    const oppositeOffset = (offset * -1) as 1 | -1;
    const edge = tile.edges.getByOffset(dimensions, dimension, offset);

    if (edge != null) {
      availableTiles = availableTiles.reduceOffset(
        edge,
        dimensions,
        dimension,
        oppositeOffset
      );
    }
    return availableTiles;
  }

  /**
   * Reduce a tileset by a neighboring tile.
   * @param tiles - The neighboring tileset
   * @param availableTiles - The tileset to reduce
   * @param dimensions - The number of dimensions
   * @param dimension - The dimension of the edge
   * @param offset - The offset from the position
   * @returns - A new tileset reduced by the neighboring tile.
   */
  private reduceByNeighborEdges(
    tiles: WangTileSet<D>,
    availableTiles: WangTileSet<D, unknown>,
    dimensions: number,
    dimension: number,
    offset: -1 | 1
  ): WangTileSet<D, unknown> {
    let beforeEdges: Edge[] | null = null;

    for (const tile of tiles.values()) {
      const beforeEdge = tile.edges.getByOffset(dimensions, dimension, offset);

      if (beforeEdge != null) {
        if (beforeEdges == null) {
          beforeEdges = [];
        }
        beforeEdges.push(beforeEdge);
      }
    }

    if (beforeEdges != null) {
      return availableTiles.reduceOffset(
        beforeEdges,
        dimensions,
        dimension,
        -1
      );
    }

    return availableTiles;
  }

  /**
   * Convert coordinate map to a simple object for storage and testing.
   * @returns An object representing all coordinates in the map. Keys are comma delineated coordinates.
   */
  toJSON(): Record<string, V> {
    return this.coordMap.toJSON();
  }
}
