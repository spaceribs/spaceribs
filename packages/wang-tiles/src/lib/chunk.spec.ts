import { WangTileChunk } from './chunk';
import { Edge } from './edge';
import { RandomGenerator } from './prng';
import { WangTile } from './tile';
import { WangTileSet } from './tile-set';
import { asciiDebug } from './util/ascii-debug';

describe('WangTileChunk', () => {
  describe('constructor', () => {
    let tiles: WangTileSet;
    let random: RandomGenerator;
    beforeEach(() => {
      const wangTile1 = new WangTile('-');
      const wangTile2 = new WangTile('|');
      tiles = new WangTileSet([wangTile1, wangTile2]);
      random = new RandomGenerator('testing');
    });

    it('should initialize a 1 dimensional chunk', () => {
      const chunk = new WangTileChunk(tiles, random, [0], [3]);
      expect(chunk.toJSON()).toMatchSnapshot();
    });

    it('should initialize a 2 dimensional chunk', () => {
      const chunk = new WangTileChunk(tiles, random, [0, 0], [3, 3]);
      expect(chunk.toJSON()).toMatchSnapshot();
    });

    it('should initialize a 3 dimensional chunk', () => {
      const chunk = new WangTileChunk(tiles, random, [0, 0, 0], [3, 3, 3]);
      expect(chunk.toJSON()).toMatchSnapshot();
    });
  });

  describe('observe()', () => {
    let chunk: WangTileChunk<[number, number]>;
    let tiles: WangTileSet<[number, number]>;
    let random: RandomGenerator;

    beforeEach(() => {
      const wangTile1 = new WangTile<[number, number]>('-');
      const wangTile2 = new WangTile<[number, number]>('|');
      tiles = new WangTileSet([wangTile1, wangTile2]);
      random = new RandomGenerator('testing');
      chunk = new WangTileChunk(tiles, random, [0, 0], [5, 10]);
    });

    it('should observe the center of the chunk first.', () => {
      const result = chunk.observe().next().value;
      expect(result).toMatchInlineSnapshot(`
        Array [
          Array [
            2,
            5,
          ],
          WangTile {
            "data": "-",
            "edges": Object {},
            "probability": 1,
          },
        ]
      `);
      expect(asciiDebug(chunk)).toMatchInlineSnapshot(`
        "
        ░░░░░
        ░░░░░
        ░░░░░
        ░░░░░
        ░░-░░
        ░░░░░
        ░░░░░
        ░░░░░
        ░░░░░
        ░░░░░
        "
      `);
    });

    it('should observe all the tiles.', () => {
      const results = [...chunk.observe()];
      expect(asciiDebug(chunk)).toMatchInlineSnapshot(`
        "
        ---|-
        ||-||
        -----
        ||-||
        ---|-
        -----
        -|-|-
        |-|--
        |||--
        |----
        "
      `);
      expect(results).toMatchSnapshot();
      expect(chunk.observed).toBe(true);
    });

    it('should observe and restrict tilesets by constraint.', () => {
      const wangTile1 = new WangTile<[number, number]>(' ');
      const wangTile2 = new WangTile<[number, number]>('-');
      const wangTile3 = new WangTile<[number, number]>('=');

      const edge1: Edge = Symbol('one');
      const edge2: Edge = Symbol('two');
      const edge3: Edge = Symbol('three');

      wangTile1.addConstraint({
        edge: edge1,
        coords: [-1, 0],
      });
      wangTile1.addConstraint({
        edge: edge2,
        coords: [1, 0],
      });

      wangTile2.addConstraint({
        edge: edge2,
        coords: [-1, 0],
      });
      wangTile2.addConstraint({
        edge: edge3,
        coords: [1, 0],
      });

      wangTile3.addConstraint({
        edge: edge3,
        coords: [-1, 0],
      });
      wangTile3.addConstraint({
        edge: edge1,
        coords: [1, 0],
      });

      tiles = new WangTileSet<[number, number]>([
        wangTile1,
        wangTile2,
        wangTile3,
      ]);

      random = new RandomGenerator('testing12');

      chunk = new WangTileChunk(tiles, random, [0, 0], [10, 5]);

      const results = [...chunk.observe()];

      expect(asciiDebug(chunk)).toMatchInlineSnapshot(`
        "
        -= -= -= -
        -= -= -= -
        = -= -= -=
        -= -= -= -
        -= -= -= -
        "
      `);
      expect(chunk.observed).toBe(true);
    });

    it('should generate contiguous rooms.', () => {
      const empty_edge: Edge = Symbol('empty_edge');
      const vert_line_edge: Edge = Symbol('vert_line_bottom_edge');
      const horiz_line_edge: Edge = Symbol('horiz_line_left_edge');

      const empty = new WangTile<[number, number]>(' ', 10);
      empty.addConstraint({ edge: empty_edge, coords: [1, 0] });
      empty.addConstraint({ edge: empty_edge, coords: [-1, 0] });
      empty.addConstraint({ edge: empty_edge, coords: [0, 1] });
      empty.addConstraint({ edge: empty_edge, coords: [0, -1] });

      const vert = new WangTile<[number, number]>('║');
      vert.addConstraint({ edge: empty_edge, coords: [1, 0] });
      vert.addConstraint({ edge: empty_edge, coords: [-1, 0] });
      vert.addConstraint({ edge: vert_line_edge, coords: [0, 1] });
      vert.addConstraint({ edge: vert_line_edge, coords: [0, -1] });

      const horiz = new WangTile<[number, number]>('═');
      horiz.addConstraint({ edge: horiz_line_edge, coords: [1, 0] });
      horiz.addConstraint({ edge: horiz_line_edge, coords: [-1, 0] });
      horiz.addConstraint({ edge: empty_edge, coords: [0, 1] });
      horiz.addConstraint({ edge: empty_edge, coords: [0, -1] });

      const topleft = new WangTile<[number, number]>('╔');
      topleft.addConstraint({ edge: horiz_line_edge, coords: [1, 0] });
      topleft.addConstraint({ edge: empty_edge, coords: [-1, 0] });
      topleft.addConstraint({ edge: empty_edge, coords: [0, 1] });
      topleft.addConstraint({ edge: vert_line_edge, coords: [0, -1] });

      const topright = new WangTile<[number, number]>('╗');
      topright.addConstraint({ edge: empty_edge, coords: [1, 0] });
      topright.addConstraint({ edge: horiz_line_edge, coords: [-1, 0] });
      topright.addConstraint({ edge: empty_edge, coords: [0, 1] });
      topright.addConstraint({ edge: vert_line_edge, coords: [0, -1] });

      const bottomleft = new WangTile<[number, number]>('╚');
      bottomleft.addConstraint({ edge: horiz_line_edge, coords: [1, 0] });
      bottomleft.addConstraint({ edge: empty_edge, coords: [-1, 0] });
      bottomleft.addConstraint({ edge: vert_line_edge, coords: [0, 1] });
      bottomleft.addConstraint({ edge: empty_edge, coords: [0, -1] });

      const bottomright = new WangTile<[number, number]>('╝');
      bottomright.addConstraint({ edge: empty_edge, coords: [1, 0] });
      bottomright.addConstraint({
        edge: horiz_line_edge,
        coords: [-1, 0],
      });
      bottomright.addConstraint({
        edge: vert_line_edge,
        coords: [0, 1],
      });
      bottomright.addConstraint({ edge: empty_edge, coords: [0, -1] });

      tiles = new WangTileSet<[number, number]>([
        empty,
        vert,
        horiz,
        topleft,
        topright,
        bottomleft,
        bottomright,
      ]);

      random = new RandomGenerator('foo');

      chunk = new WangTileChunk(tiles, random, [0, 0], [24, 24]);

      const results = [...chunk.observe()];

      expect(asciiDebug(chunk)).toMatchInlineSnapshot(`
        "
           ║ ║     ╚══╗║  ╚╝    
           ║╔╝        ║║     ╔══
           ╚╝   ╔══╗  ╚╝ ╔══╗║  
          ╔╗    ╚══╝     ║  ╚╝  
         ╔╝║             ╚═══╗  
         ╚═╝╔═╗   ╔╗    ╔══╗ ║  
        ╗   ║ ║   ╚╝    ╚══╝ ╚╗ 
        ║ ╔═╝╔╝ ╔══╗          ║ 
        ║ ║  ║  ╚╗ ║          ╚═
        ║ ╚══╝╔══╝ ║   ╔╗       
        ╝    ╔╝  ╔═╝   ╚╝╔╗     
             ║   ╚══╗╔╗  ╚╝╔═╗  
        ╗    ║      ╚╝║  ╔╗║ ╚╗ 
        ╝    ╚═╗    ╔╗╚╗ ╚╝║ ╔╝ 
               ║    ║╚═╝   ║ ║  
              ╔╝    ╚══╗   ║ ╚╗ 
         ╔╗   ╚╗      ╔╝   ╚╗ ╚═
        ╗║║    ╚╗ ╔══╗║     ╚╗╔╗
        ╝║║    ╔╝ ║  ╚╝      ║║╚
         ║╚═╗  ║  ║          ╚╝ 
         ╚╗ ╚══╝╔╗╚╗            
          ╚═╗   ║╚╗║ ╔╗ ╔╗   ╔══
        ╗   ╚═╗╔╝╔╝║ ╚╝ ╚╝   ╚╗ 
        ╚═╗   ║║╔╝ ║╔╗ ╔═══╗ ╔╝ 
        "
      `);
      expect(chunk.observed).toBe(true);
    });
  });
});
