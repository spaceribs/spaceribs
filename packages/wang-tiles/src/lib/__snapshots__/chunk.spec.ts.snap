// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WangTileChunk constructor should initialize a 1 dimensional chunk 1`] = `
Object {
  "0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
}
`;

exports[`WangTileChunk constructor should initialize a 2 dimensional chunk 1`] = `
Object {
  "0,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
}
`;

exports[`WangTileChunk constructor should initialize a 3 dimensional chunk 1`] = `
Object {
  "0,0,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,0,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,0,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,1,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,1,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,1,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,2,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,2,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "0,2,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,0,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,0,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,0,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,1,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,1,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,1,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,2,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,2,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "1,2,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,0,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,0,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,0,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,1,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,1,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,1,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,2,0": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,2,1": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
  "2,2,2": Set {
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  },
}
`;

exports[`WangTileChunk observe() should observe all the tiles. 2`] = `
Array [
  Array [
    Array [
      2,
      5,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      5,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      4,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      4,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      3,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      3,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      2,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      2,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      1,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      1,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      6,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      6,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      7,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      7,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      8,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      8,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      9,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      9,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      9,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      8,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      7,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      6,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      5,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      4,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      3,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      2,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      1,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      9,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      8,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      7,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      6,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      5,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      4,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      3,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      2,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      1,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      9,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      8,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      7,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      6,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      5,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      4,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      3,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      2,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      1,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      4,
      0,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      3,
      0,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      2,
      0,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      1,
      0,
    ],
    WangTile {
      "data": "-",
      "edges": Object {},
      "probability": 1,
    },
  ],
  Array [
    Array [
      0,
      0,
    ],
    WangTile {
      "data": "|",
      "edges": Object {},
      "probability": 1,
    },
  ],
]
`;
